src/lib.rs

```diff
@@ -1044,6 +1044,153 @@ where
+                }
+                BinEntry::Node(ref head) => {
+                    let head_lock = head.lock.lock();
+                    let mut old_val: Option<Shared<'g, V>> = None;
+
+                    // need to check that this is _still_ the head
+                    if t.bin(i, guard) != bin {
+                        continue;
+                    }
+
+                    // TODO: tree nodes
+                    let mut e = bin;
+                    let mut pred: Shared<'_, BinEntry<K, V>> = Shared::null();
+                    loop {
+                        // safety: either e is bin, in which case it is valid due to the above,
+                        // or e was obtained from a next pointer. Any next pointer obtained from
+                        // bin is valid at the time we look up bin in the table, at which point
+                        // the epoch is pinned by our guard. Since we found the next pointer
+                        // in a valid map and it is not null (as checked above and below), the
+                        // node it points to was present (i.e. not removed) from the map in the
+                        // current epoch. Thus, because the epoch cannot advance until we release
+                        // our guard, e is also valid if it was obtained from a next pointer.
+                        let n = unsafe { e.deref() }.as_node().unwrap();
+                        let next = n.next.load(Ordering::SeqCst, guard);
+                        if n.hash == h && &n.key == key {
+                            let ev = n.value.load(Ordering::SeqCst, guard);
+                            old_val = Some(ev);
+
+                            // remove the BinEntry containing the removed key value pair from the bucket
+                            if !pred.is_null() {
+                                // either by changing the pointer of the previous BinEntry, if present
+                                // safety: as above
+                                unsafe { pred.deref() }
+                                    .as_node()
+                                    .unwrap()
+                                    .next
+                                    .store(next, Ordering::SeqCst);
+                            } else {
+                                // or by setting the next node as the first BinEntry if there is no previous entry
+                                t.store_bin(i, next);
+                            }
+
+                            // in either case, mark the BinEntry as garbage, since it was just removed
+                            // safety: as for val below / in put
+                            unsafe { guard.defer_destroy(e) };
+
+                            // since the key was found and only one node exists per key, we can
+                            // break here (as an addition to the Java code)
+                            break;
```

Comment:
The Java code also does this at
https://github.com/jonhoo/flurry/blob/8c014c349df1cbe03f98d84f7d59115899e4df7f/jsr166/src/ConcurrentHashMap.java#L1113

So I'm not sure what the comment is referring to?